config/config.go

```diff
@@ -370,19 +376,99 @@ var defaultConf = Config{
 	},
 }
 
-var globalConf = defaultConf
+var (
+	globalConf             = atomic.Value{}
+	reloadConfPath         = ""
+	confReloader           func(nc, c *Config)
+	confReloadLock         sync.Mutex
+	supportedReloadConfigs = make(map[string]struct{}, 32)
+)
 
 // NewConfig creates a new config instance with default value.
 func NewConfig() *Config {
 	conf := defaultConf
 	return &conf
 }
 
+// SetConfReloader sets reload config path and a reloader.
+// It should be called only once at start time.
+func SetConfReloader(cpath string, reloader func(nc, c *Config), confItems ...string) {
+	reloadConfPath = cpath
+	confReloader = reloader
+	for _, item := range confItems {
+		supportedReloadConfigs[item] = struct{}{}
+	}
+}
+
 // GetGlobalConfig returns the global configuration for this server.
 // It should store configuration from command line and configuration file.
 // Other parts of the system can read the global configuration use this function.
 func GetGlobalConfig() *Config {
-	return &globalConf
+	return globalConf.Load().(*Config)
+}
+
+// ReloadGlobalConfig reloads global configuration for this server.
+func ReloadGlobalConfig() error {
+	confReloadLock.Lock()
+	defer confReloadLock.Unlock()
+
+	nc := NewConfig()
+	if err := nc.Load(reloadConfPath); err != nil {
+		return err
+	}
+	if err := nc.Valid(); err != nil {
+		return err
+	}
+	c := GetGlobalConfig()
+
+	diffs := collectsDiff(*nc, *c, "")
+	if len(diffs) == 0 {
+		return nil
+	}
+	for k := range diffs {
+		if _, ok := supportedReloadConfigs[k]; !ok {
+			return fmt.Errorf("reloading config %s is not supported", k)
```

Comment:
It is better to show some info, e.g. "[config_name] should be [config_value]", to help the user to rollback their modification.

Reply:
updated.